feat(flags): feature flags endpoint

lzap opened the pull request:
https://issues.redhat.com/browse/HMSPROV-314

Implements feature flag endpoint that our/ib UI will use to show/hide the launch button. The backend service we use is named Unleash

* https://www.getunleash.io/
* https://github.com/Unleash/unleash
* https://github.com/Unleash/unleash-client-go

Therefore all the configuration variables are also named `UNLEASH_` and it is possible to configure the service locally and enable it for debugging purposes. When Unleash is disabled through config, all queries for feature flags return true. When running in Clowder, Unleash gets enabled by default and configured from Clowder.

ezr-ondrej reviewed (approved):
Awesome work :heart_eyes:

ezr-ondrej commented:
I'll keep it up in case there are others who wants to take a look, but I'm all for merging right away :)

adiabramovitch reviewed (commented):
LGTM, just a small question:
will ib will use this endpoint in order to know if the launch instances button should be visible or not?

lzap commented:
> will ib will use this endpoint in order to know if the launch instances button should be visible or not?

That was the plan and the reason why we actually expose this via an endpoint.

@ezr-ondrej before merging, I think we should put this into OpenAPI. I am not sure.

adiabramovitch commented:
I think it needs to be in the OpenAPI

ezr-ondrej commented on internal/routes/all_routes.go:
```diff
@@ -114,6 +114,10 @@ func apiRouter() http.Handler {
 			})
 		})
+
+		r.Route("/feature_flag/{FLAG}", func(r chi.Router) {
```
I'd probably not expose it for now, it seems that we can consume other endpoints in the UI, that are already exposed for us, backend should consume feature flags to make it's own decisions only if I understand it correctly.

lzap replied:
Well I thought so but you wrote this into the ticket. IMHO both UI and API should have its own unleash code checking against the server, no endpoint would ever be needed.

ezr-ondrej replied:
yeah, you're right, I was wrong. Let me update the ticket.

ezr-ondrej commented on internal/config/flags.go:
```diff
@@ -0,0 +1,62 @@
+		return true
+	}
+
+	return unleash.IsEnabled(name)
```
Also, we shall probably pass a context with some user info, so we can use the rollout strategies (like A-B testing).

https://github.com/Unleash/unleash-client-go#unleash-context

lzap replied:
Yeah I did not know what strategies are used on our instances, so I will simply use orgID for the "userId" in the context for the initial version.

ezr-ondrej reviewed (commented):
few thoughts on improvements, but no blockers ^_^
I'd say we do not need to expose it at all, but if we do we don't need to expose it in the OpenAPI, as this is not customer facing endpoint and we don't want to have QE tests for it IMO :)

adiabramovitch commented:
> few thoughts on improvements, but no blockers ^_^
> 
> I'd say we do not need to expose it at all, but if we do we don't need to expose it in the OpenAPI, as this is not customer facing endpoint and we don't want to have QE tests for it IMO :)

I am not sure why , I would assume we want we to test when the bottom appears and under which circumstances

lzap commented on internal/config/flags.go:
```diff
@@ -0,0 +1,62 @@
+
+func FeatureEnabled(name string) bool {
+	// When unleash is not configured, all features are enabled by default.
+	if !config.Unleash.Enabled {
+		return true
+	}
+
+	return unleash.IsEnabled(name)
```
I do prefer more readable style at cost of LOC but you know what let's do that so I look more clever :-)

avitova replied:
If it feels less readable, let's not force it to be there.:)

lzap replied:
Aah after I implemented contexts, it is too messy so I had to fallback to the if statement at the beginning.

ezr-ondrej commented:
> I am not sure why , I would assume we want we to test when the bottom appears and under which circumstances

I'm not sure there, @RHEnVision/qe what would you say? :)

lzap commented:
Well you gave an answer above, @ezr-ondrej - you said that we should not expose anything at all. Therefore no OpenAPI is needed.

In general, think about OpenAPI as a huge commitment. We should only add things that we are 100% sure to have there for years and when they are stable. I am not so sure about this one.

lzap commented:
Rebased, implemented user context. I am going for org_id for userId mapping for now, if we have our own instance that is exactly what we need. If we are going to use a shared one, then I can change this if needed.

I slightly refactored the endpoint, it is still undocumented via OpenAPI, it is more simple now tho - it exposes any flag that starts with "app.provisioning". I am not adding this into OpenAPI because chances are that we might be sharing a single unleash instance with image builder, and in that case no such endpoint would be necessary.

Please test this before merging :-)

lzap commented on internal/config/config.go:
```diff
@@ -107,6 +107,13 @@ var config struct {
+	Unleash struct {
+		Enabled     bool   `env:"ENABLED" env-default:"false" env-description:"unleash service (feature flags)"`
+		Environment string `env:"ENVIRONMENT" env-default:"" env-description:"unleash environment"`
+		Prefix      string `env:"PREFIX" env-default:"app.provisioning" env-description:"unleash flag prefix"`
```
Also environment is configurable option now, this will get used if we will have a single instance across environments, which is not the case I think. But I added it, just do not set it and an empty string is the default value that should not hurt anything in the client.

lzap commented on internal/services/reservations_service.go:
```diff
@@ -20,6 +21,10 @@ var (
+
 	pType := models.ProviderTypeFromString(chi.URLParam(r, "TYPE"))
 	switch pType {
 	case models.ProviderTypeNoop:
```
This is the place - when launching is disabled for an account or globally, then it returns unauthorized. Listing reservations is possible (through API) tho.

lzap commented on internal/config/config.go:
```diff
@@ -107,6 +107,13 @@ var config struct {
+		Prefix      string `env:"PREFIX" env-default:"app.provisioning" env-description:"unleash flag prefix"`
+		URL         string `env:"URL" env-default:"http://localhost:4242" env-description:"unleash service URL"`
+		Token       string `env:"TOKEN" env-default:"" env-description:"unleash service client access token"`
+	} `env-prefix:"UNLEASH_"`
```
I also added environment, by default it is unused and we will likely not need it (only if we share a single instance across environments which is not the case I think).

lzap commented on internal/services/error_renderer.go:
```diff
@@ -35,3 +46,33 @@ func renderNotFoundOrDAOError(w http.ResponseWriter, r *http.Request, err error,
+
+func writeUnauthorized(w http.ResponseWriter, r *http.Request) {
+	writeEmptyResponse(w, r, http.StatusUnauthorized)
+}
```
I moved these function from various places into a single place and renamed them so it`s more readable.

lzap commented:
Ok so now there's an error UNAUTHORIZED during AWS EC2 launch, that's because Unleash on ephemeral has no data, thus the flag is FALSE. How do we solve that? :-) @ezr-ondrej

ezr-ondrej commented:
Can we use `WithFallback` to return true on missing toggle?

https://github.com/Unleash/unleash-client-go/blob/459ff5b251921855426f98ad1e7091df21a37e23/config.go#L167

lzap commented:
Good idea, rebased.

akhil-jha commented:
/retest

lzap commented:
Rebased.

lzap commented:
All good, please re-review since I had bunch of merge conflicts, thanks!

ezr-ondrej reviewed (approved):
Nice, I'm not sure if the stage environment will pass the token correctly through clowder, but there is one way to find out I guess ;)
